### PR TITLE
Add JITServer conditional build flag

### PR DESCRIPTION
**Pull request by a7ehuo**
Add ` JITServer` conditional build flag `—enable-jitserver`. 

Adding `—enable-jitserver` or setting `—enable-jitserver=yes` in `bash configure` for target OS `linux` exports `JITSERVER_SUPPORT=1` to the JIT build environment.
    
Either one of the following settings will not build the `JITServer`:
- Setting `—enable-jitserver=yes` for non-linux OS.
- Without specifying the `—enable-jitserver` flag.
- Explicitly setting `—enable-jitserver=no`.

If building `JITServer`, the installed protobuf version will be checked:
- For `x86`, protobuf `3.5.1` or above is required.
- For other platforms, protobuf `3.7.1` or above is required.

For example to build JIT with `JITServer` support: 
```
bash configure --with-freemarker-jar=/root/freemarker.jar --with-boot-jdk=/root/boot/jdk8 --enable-jitserver=yes
make all
```

This is the new PR that is related to what #315 was trying to do. 

Signed-off-by: Annabelle Huo <Annabelle.Huo@ibm.com>

### CODE REVIEW

File: jdk/make/closed/autoconf/custom-hook.m4 [hunk 1]
```diff
@@ -190,6 +191,31 @@ AC_DEFUN([OPENJ9_CONFIGURE_DDR],
...
+  AC_MSG_CHECKING([for jitserver])
+  AC_ARG_ENABLE([jitserver], [AS_HELP_STRING([--enable-jitserver], [enable JITServer support @<:@disabled@:>@])])
+  if test "x$enable_jitserver" = xyes ; then
+    if test "x$OPENJDK_TARGET_OS" = xlinux ; then
+        AC_MSG_RESULT([yes (explicitly enabled)])
+        OPENJ9_ENABLE_JITSERVER=true
+    else
+        AC_MSG_RESULT([no (default for $OPENJDK_TARGET_OS)])
+        OPENJ9_ENABLE_JITSERVER=false
+    fi
```
Comment (keithc-ca) [inline jdk/make/closed/autoconf/custom-hook.m4]: Please indent consistently (two spaces per level).

File: closed/OpenJ9.gmk [hunk 2]
```diff
@@ -408,6 +408,9 @@ endif
 ifneq (,$(OPENJ9_DEVELOPER_DIR))
   CUSTOM_COMPILER_ENV_VARS += DEVELOPER_DIR="$(OPENJ9_DEVELOPER_DIR)"
 endif
+ifeq (true,$(OPENJ9_ENABLE_JITSERVER))
+  CUSTOM_COMPILER_ENV_VARS += JITSERVER_SUPPORT=1
+endif
```
Comment (keithc-ca) [inline closed/OpenJ9.gmk]: I'm not thrilled about the use of environment variables to communicate configuration options, but in the cmake case, I believe this is redundant.

File: closed/OpenJ9.gmk [hunk 3]
```diff
@@ -463,6 +466,11 @@ ifneq (,$(CCACHE))
   CMAKE_ARGS += "-DCMAKE_CXX_COMPILER_LAUNCHER=$(ESCAPED_CCACHE)"
   CMAKE_ARGS += "-DCMAKE_C_COMPILER_LAUNCHER=$(ESCAPED_CCACHE)"
 endif # CCACHE
+ifeq (true,$(OPENJ9_ENABLE_JITSERVER))
+  CMAKE_ARGS += "-DJITSERVER_SUPPORT=ON"
+else
+  CMAKE_ARGS += "-DJITSERVER_SUPPORT=OFF"
```
Comment (keithc-ca) [inline closed/OpenJ9.gmk]: Please remove the unnecessary double-quotes (`"`).

**Comment by keithc-ca** [comment]
Use of `--enable-jitserver` should verify that required packages (e.g. `protobuf`) are available.

File: closed/OpenJ9.gmk (hunk 2 above)
Comment (a7ehuo) [reply closed/OpenJ9.gmk]: `JITSERVER_SUPPORT` flag will eventually be removed once `JITServer` is built by default.
cmake will check `JITSERVER_SUPPORT` as a variable definition.

Comment (keithc-ca) [reply closed/OpenJ9.gmk]: Please arrange to set `JITSERVER_SUPPORT` only where necessary (i.e. *not* for cmake).

**Comment by keithc-ca** [comment]
My first attempt to build with this new option failed due to mis-matched PIC settings; https://github.com/eclipse/openj9/pull/7767 fixes the broken Dockerfiles I used as a guide.

Even after fixing that, building with cmake yields this error:
```
CMake Error at runtime/compiler/CMakeLists.txt:92 (string):
  string sub-command REPLACE requires at least four arguments.
```

**Comment by a7ehuo** [comment]
@keithc-ca I'm trying to reproduce the error from `runtime/compiler/CMakeLists.txt`. 
1c98d23 only addresses the code review feedback. 
@mpirvu FYI.

File: jdk/make/closed/autoconf/custom-hook.m4 [hunk 4]
```diff
@@ -212,6 +213,55 @@ AC_DEFUN([OPENJ9_PLATFORM_EXTRACT_VARS_FROM_CPU],
...
+          OPENJ9_ENABLE_JITSERVER=true
+        else
+          AC_MSG_ERROR([Require protobuf >= ($MIN_SUPPORTED_PROTOBUF_VERSION) for ($OPENJ9_CPU)])
+        fi
+
+      else
+        AC_MSG_ERROR([protoc not found])
+      fi
+    else
+      AC_MSG_RESULT([no (default for $OPENJDK_TARGET_OS)])
```
Comment (keithc-ca) [inline jdk/make/closed/autoconf/custom-hook.m4]: This isn't the right message if `--enable-jitserver` was specified: it should say unsupported or something similar.

Comment (keithc-ca) [reply jdk/make/closed/autoconf/custom-hook.m4]: Inverting these `if` statements would put the conditions closer to *both* branches making them easier to understand.

**Comment by a7ehuo** [comment]
@keithc-ca Thanks for finding and root causing the error in `runtime/compiler/CMakeLists.txt`. The fix is in eclipse/openj9#7789. 

All comments addressed. Ready for another review. Thanks.

File: jdk/make/closed/autoconf/custom-hook.m4 [hunk 5]
```diff
@@ -212,6 +213,55 @@ AC_DEFUN([OPENJ9_PLATFORM_EXTRACT_VARS_FROM_CPU],
...
+  if test "x$enable_jitserver" = xyes ; then
+
+    if test "x$OPENJDK_TARGET_OS" != xlinux ; then
+      AC_MSG_RESULT([no (unsupported for $OPENJDK_TARGET_OS)])
+    else
+      AC_CHECK_PROG(PROTOC_INSTALLED,protoc,yes,no)
+      if test "x$PROTOC_INSTALLED" = xno ; then
+        AC_MSG_ERROR([protoc not found])
+      else
+        AC_MSG_CHECKING([for protobuf version])
```
Comment (keithc-ca) [inline jdk/make/closed/autoconf/custom-hook.m4]: We haven't given the `AC_MSG_RESULT` that matches `AC_MSG_CHECKING([for jitserver])` above yet; the log will be a little difficult to follow.

**Comment by a7ehuo** [comment]
> Can you put together similar changes for jdk11, jdk13 and jdknext?

@keithc-ca `JITServer` is supported in `jdk11`. I'll create one for it. For now, I'm not sure if it'd be supported for `jdk13` and `jdknext` as well. 

@mpirvu Would this change need to be in `jdk13` and `jdknext`?

**Comment by mstoodle** [comment]
@a7ehuo I think the option should be there even if we choose to document that it isn't yet supported in those releases (or to document that it's only supported for JDK8 and JDK11).

**Comment by mpirvu** [comment]
I totally agree with @mstoodle

File: jdk/make/closed/autoconf/custom-spec.gmk.in [hunk 6]
```diff
@@ -48,6 +48,9 @@ WITH_OPENSSL            := @WITH_OPENSSL@
 ## DDR
 OPENJ9_ENABLE_DDR       := @OPENJ9_ENABLE_DDR@
 
+## JITSERVER
+OPENJ9_ENABLE_JITSERVER    := @OPENJ9_ENABLE_JITSERVER@
```
Comment (keithc-ca) [inline jdk/make/closed/autoconf/custom-spec.gmk.in]: Can you please remove the extra `#` here (and above for DDR), and remove the extra spaces before `:=`.

**Comment by a7ehuo** [comment]
@keithc-ca All comments addressed. Ready for another review. Thanks.

**Comment by keithc-ca** [comment]
Jenkins compile xlinux,xlinux_cm jdk8

**Comment by keithc-ca** [comment]
Jenkins compile xlinux,xlinuxcm jdk8

**Comment by keithc-ca** [comment]
Both builds failed similarly, but long after any changes here would have made a difference.
```
Creating placeholder flownodes because failed loading originals.
java.io.IOException: Tried to load head FlowNodes for execution Owner[Build_JDK8_x86-64_linux_Personal/543:Build_JDK8_x86-64_linux_Personal #543] but FlowNode was not found in storage for head id:FlowNodeId 1:110
	at org.jenkinsci.plugins.workflow.cps.CpsFlowExecution.initializeStorage(CpsFlowExecution.java:679)
	at org.jenkinsci.plugins.workflow.cps.CpsFlowExecution.onLoad(CpsFlowExecution.java:716)
	at org.jenkinsci.plugins.workflow.job.WorkflowRun.getExecution(WorkflowRun.java:665)
	at org.jenkinsci.plugins.workflow.job.WorkflowRun.onLoad(WorkflowRun.java:529)
	at hudson.model.RunMap.retrieve(RunMap.java:225)
	at hudson.model.RunMap.retrieve(RunMap.java:57)
	at jenkins.model.lazy.AbstractLazyLoadRunMap.load(AbstractLazyLoadRunMap.java:501)
	at jenkins.model.lazy.AbstractLazyLoadRunMap.load(AbstractLazyLoadRunMap.java:483)
	at jenkins.model.lazy.AbstractLazyLoadRunMap.getByNumber(AbstractLazyLoadRunMap.java:381)
	at hudson.model.RunMap.getById(RunMap.java:205)
	at org.jenkinsci.plugins.workflow.job.WorkflowRun$Owner.run(WorkflowRun.java:902)
	at org.jenkinsci.plugins.workflow.job.WorkflowRun$Owner.get(WorkflowRun.java:913)
	at org.jenkinsci.plugins.workflow.flow.FlowExecutionList$1.computeNext(FlowExecutionList.java:65)
	at org.jenkinsci.plugins.workflow.flow.FlowExecutionList$1.computeNext(FlowExecutionList.java:57)
	at com.google.common.collect.AbstractIterator.tryToComputeNext(AbstractIterator.java:143)
	at com.google.common.collect.AbstractIterator.hasNext(AbstractIterator.java:138)
	at org.jenkinsci.plugins.workflow.flow.FlowExecutionList$ItemListenerImpl.onLoaded(FlowExecutionList.java:178)
	at jenkins.model.Jenkins.<init>(Jenkins.java:1022)
	at hudson.model.Hudson.<init>(Hudson.java:85)
	at hudson.model.Hudson.<init>(Hudson.java:81)
	at hudson.WebAppMain$3.run(WebAppMain.java:233)
```
I successfully built an earlier version of this locally with eclipse/openj9#7789 using both autotools and cmake. Approving and merging.

**Pull request merged**